src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp

```diff
@@ -345,6 +346,58 @@ SIMPLE_MATCHER_PASS_DEFINITION(EliminateGather,
                                ov::op::v7::Gather,
                                ov::op::v8::Gather);
 
+pass::EliminateReduceReshape::EliminateReduceReshape() {
+    MATCHER_SCOPE(EliminateReduceReshape);
+    using namespace pass::pattern;
+    auto axes = wrap_type<ov::op::v0::Constant>();
+    auto reduce_pattern = wrap_type<ov::op::util::ArithmeticReductionKeepDims>({any_input(), axes});
+    auto requested_shape_pattern = wrap_type<ov::op::v0::Constant>();
+    auto reshape_pattern =
+        wrap_type<ov::op::v1::Reshape>({reduce_pattern, requested_shape_pattern}, consumers_count(1));
+
+    matcher_pass_callback callback = [=](Matcher& m) {
+        auto pattern_map = m.get_pattern_map();
+        auto reshape_node = m.get_match_root();
+        auto reduce_node = pattern_map.at(reduce_pattern);
+
+        auto reshape = ov::as_type_ptr<ov::op::v1::Reshape>(reshape_node);
+        auto reduce = ov::as_type_ptr<ov::op::util::ArithmeticReductionKeepDims>(reduce_node);
+        if (!reshape || !reduce) {
+            return false;
+        }
+
+        auto in_rank = reshape->get_input_partial_shape(0).rank();
+        auto out_rank = reshape->get_output_partial_shape(0).rank();
+        if (in_rank.is_dynamic() || out_rank.is_dynamic() || in_rank.get_length() != out_rank.get_length()) {
+            return false;
+        }
+
+        auto requested_shape = ov::as_type_ptr<ov::op::v0::Constant>(pattern_map.at(requested_shape_pattern));
+        if (!requested_shape) {
+            return false;
+        }
+
+        auto requested_shape_vec = requested_shape->cast_vector<int64_t>();
+        auto axes = reduce->get_reduction_axes();
+
+        int cnt_dyn = 0;
+        for (size_t i = 0; i < requested_shape_vec.size(); ++i) {
+            // if we use reshape special zero or this dim was reduced
+            cnt_dyn += !((requested_shape_vec[i] == 0 && reshape->get_special_zero()) ||
+                         (axes.count(i) && requested_shape_vec[i] == 1));
+        }
+
+        if (cnt_dyn <= 1) {
+            return replace_output_update_name(reshape->output(0), reshape->input_value(0));
+        } else {
+            return false;
+        }
```

Comment:
What are the conditions you're checking against?

Reply:
added a comment